models/events.go

```diff
@@ -1,114 +1,142 @@
 func isBlockExecutedEvent(event cadence.Event) bool {
 	if event.EventType == nil {
 		return false
 	}
-	return strings.Contains(event.EventType.ID(), string(types.EventTypeBlockExecuted))
+	return strings.Contains(event.EventType.ID(), string(events.EventTypeBlockExecuted))
 }
 
 // isTransactionExecutedEvent checks whether the given event contains transaction executed data.
 func isTransactionExecutedEvent(event cadence.Event) bool {
 	if event.EventType == nil {
 		return false
 	}
-	return strings.Contains(event.EventType.ID(), string(types.EventTypeTransactionExecuted))
+	return strings.Contains(event.EventType.ID(), string(events.EventTypeTransactionExecuted))
 }
 
+// CadenceEvents contains Flow emitted events containing one or zero evm block executed event,
+// and multiple or zero evm transaction events.
 type CadenceEvents struct {
-	events flow.BlockEvents
+	events       flow.BlockEvents  // Flow events for a specific flow block
+	block        *Block            // EVM block (at most one per Flow block)
+	transactions []Transaction     // transactions in the EVM block
+	receipts     []*StorageReceipt // receipts for transactions
 }
 
-func NewCadenceEvents(events flow.BlockEvents) *CadenceEvents {
-	return &CadenceEvents{events: events}
-}
+// NewCadenceEvents decodes the events into evm types.
+func NewCadenceEvents(events flow.BlockEvents) (*CadenceEvents, error) {
+	e, err := decodeCadenceEvents(events)
+	if err != nil {
+		return nil, err
+	}
 
-// Blocks finds the block evm events and decodes it into the blocks slice,
-// if no block events are found nil slice is returned.
-//
-// Return values:
-// blocks, nil - if blocks are found
-// nil, nil - if no block are found
-// nil, err - unexpected error
-func (c *CadenceEvents) Blocks() ([]*types.Block, error) {
-	blocks := make([]*types.Block, 0)
-	for _, e := range c.events.Events {
-		if isBlockExecutedEvent(e.Value) {
-			block, err := decodeBlock(e.Value)
-			if err != nil {
-				return nil, err
-			}
-			blocks = append(blocks, block)
-		}
+	// if cadence event is empty don't calculate any dynamic values
+	if e.Empty() {
+		return e, nil
 	}
 
-	// make sure order of heights is ordered, this is safety check
-	slices.SortFunc(blocks, func(a, b *types.Block) int {
-		return int(a.Height - b.Height)
-	})
+	// calculate dynamic values
+	cumulativeGasUsed := uint64(0)
+	blockHash, err := e.block.Hash()
+	if err != nil {
+		return nil, err
+	}
 
-	return blocks, nil
+	for i, rcp := range e.receipts {
+		// add transaction hashes to the block
+		e.block.TransactionHashes = append(e.block.TransactionHashes, rcp.TxHash)
+		// calculate cumulative gas used up to that point
+		cumulativeGasUsed += rcp.GasUsed
+		rcp.CumulativeGasUsed = cumulativeGasUsed
+		// set the transaction index
+		rcp.TransactionIndex = uint(i)
+		// set calculate block hash
+		rcp.BlockHash = blockHash
+		// dynamically add missing log fields
+		for _, l := range rcp.Logs {
```

Comment:
```suggestion
for _, l := range rcp.Logs {
	l.BlockNumber = rcp.BlockNumber.Uint64()
	l.BlockHash = rcp.BlockHash
	l.TxHash = rcp.TxHash
	l.TxIndex = rcp.TransactionIndex
	l.Index = rcp.TransactionIndex
	l.Removed = false
}
```

The consensus fields of `Log` type are only `Address` / `Topics` / `Data`, which means we have to dynamically add the rest. See: https://github.com/onflow/go-ethereum/blob/master/core/types/log.go#L30-L36